File: src/main/java/spring/springserver/global/config/MailConfig.java
package spring.springserver.global.config;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.JavaMailSenderImpl;

import java.util.Properties;

@Configuration
public class MailConfig {

	@Value("${spring.mail.username}")
	private String username;
	@Value("${spring.mail.password}")
	private String password;
	@Value("${spring.mail.host}")
	private String host;

	@Bean
	public JavaMailSender mailSender() {
		JavaMailSenderImpl mailSender = new JavaMailSenderImpl();

		mailSender.setHost(host);
		mailSender.setPort(587);
		mailSender.setUsername(username);
		mailSender.setPassword(password);

		Properties mailProperties = new Properties();
		mailProperties.put("mail.transport.protocol", "smtp");
		mailProperties.put("mail.debug", "true");
		mailProperties.put("mail.smtp.ssl.trust", "smtp.gmail.com");
		mailProperties.put("mail.smtp.ssl.protocols", "TLSv1.3");
		mailProperties.put("mail.smtp.auth", "true");
		mailProperties.put("mail.smtp.starttls.enable", "true");

		mailSender.setJavaMailProperties(mailProperties);
		return mailSender;
	}
}

File: src/main/java/spring/springserver/global/config/mail/MailConfig.kt
package spring.springserver.global.config.mail

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.JavaMailSenderImpl
import org.thymeleaf.TemplateEngine
import spring.springserver.domain.email.service.impl.EmailServiceImpl
import spring.springserver.global.config.redis.RedisConfig
import java.util.Properties

@Configuration
class MailConfig(
    @Value("\${spring.mail.username}")
    private val username: String,

    @Value("\${spring.mail.password}")
    private val password: String,

    @Value("\${spring.mail.host}")
    private val host: String
) {

    @Bean
    fun emailService(
        javaMailSender: JavaMailSender,
        redisConfig: RedisConfig,
        templateEngine: TemplateEngine
    ): EmailServiceImpl {

        return EmailServiceImpl(
            javaMailSender,
            redisConfig,
            templateEngine
        )
    }

    @Bean
    fun mailSender(): JavaMailSender {

        return JavaMailSenderImpl().apply {

            host = this@MailConfig.host
            port = 587
            username = this@MailConfig.username
            password = this@MailConfig.password
            javaMailProperties = Properties().apply {

                put("mail.transport.protocol", "smtp")
                put("mail.debug", "true")
                put("mail.smtp.ssl.trust", "smtp.gmail.com")
                put("mail.smtp.ssl.protocols", "TLSv1.3")
                put("mail.smtp.auth", "true")
                put("mail.smtp.starttls.enable", "true")
            }
        }
    }
}
